src/main/java/com/google/sps/servlets/PotentialMatchesServlet.java

```diff
@@ -60,28 +70,69 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
+  * already in datastore.
   *
-  * <p>Currently is just loading all the possible matches and obtaining one of them.
+  * <p>Creates new match information entity for user if it had not already been stored.
   *
-  * <p>TODO(#21): Obtain potential match results from datastore instead of running the 
-  * potential match finding methods repeatedly.
+  * @param userID The ID of the user who's match information is being checked/loaded into datastore
+  */
+  private void loadUserMatchInformation(String userID) {
+    Entity matchInfoEntity = datastore.prepare(new Query(MATCH_INFO_ENTITY).setFilter(
+      new FilterPredicate(USER_ID_PROPERTY, FilterOperator.EQUAL, userID))).asSingleEntity();
+    
+    if (matchInfoEntity == null) {
+      addMatchInfoToDatastore(userID);
+    }
+  }
+  
+  /**
+  * Given a specific user, get their next potential match for their feed page
+  *
+  * @param userID The user whose potential match is being retrieved
+  */
+  private String getNextPotentialMatchID(String userID) {
+    Entity matchInfoEntity = datastore.prepare(new Query(MATCH_INFO_ENTITY).setFilter(
+      new FilterPredicate(USER_ID_PROPERTY, FilterOperator.EQUAL, userID))).asSingleEntity();
+
+    List<String> potentialMatches = (ArrayList<String>) matchInfoEntity.getProperty(POTENTIAL_MATCHES_PROPERTY);
+
+    String nextPotentialMatchID;
+    
+    if (potentialMatches != null && potentialMatches.iterator().hasNext()) {
+      nextPotentialMatchID = potentialMatches.iterator().next();
+
+      potentialMatches.remove(nextPotentialMatchID);
+
+      matchInfoEntity.setProperty(POTENTIAL_MATCHES_PROPERTY, potentialMatches);
+      datastore.put(matchInfoEntity);
+    } else {
+      nextPotentialMatchID = NO_POTENTIAL_MATCH_RESULT;
+    }
+    return nextPotentialMatchID;
+  }
+
+  /**
+  * Creates a new match-info entity and adds it to the datastore
   *
-  * @param userID The userID of the user who's potential match is being found
-  * @return The next potential match for a user
+  * @param userID the user ID of the user whose match information is being stored
   */
-  private String loadUserPotentialMatch(String userID) {
+  private void addMatchInfoToDatastore(String userID) {
+    //Initialize user nodes and friend map
     ImmutableSet<UserNode> userNodes = createUserNodes();
     UserFriendsMap friendsMap = new UserFriendsMap(userNodes);
     
+    //Run the potential matching algorithm to find all potential matches
     ImmutableSet<String> potentialMatches = PotentialMatchAlgorithm.findPotentialMatchesForUser(userID, friendsMap);
-  
-    String nextPotentialMatchID = potentialMatches.iterator().hasNext() ?
-      potentialMatches.iterator().next() : NO_POTENTIAL_MATCH_RESULT;
+
+    //Add new Match Info entity to datastore
+    Entity newMatchInfo = new Entity(MATCH_INFO_ENTITY);
+    newMatchInfo.setProperty(USER_ID_PROPERTY, userID);
+    newMatchInfo.setProperty(POTENTIAL_MATCHES_PROPERTY, new ArrayList<String>(potentialMatches));
```

Comment:
```suggestion
    newMatchInfo.setProperty(POTENTIAL_MATCHES_PROPERTY, ImmutableList.copyOf(potentialMatches));
```

(prefer immutable types wherever you can)

Reply:
Got it!